File: langs/typescript/src/components/componentVMOf.ts
/**
 * ComponentVMOf<M> — modeled leaf viewmodel with settable model.
 *
 * See spec/05-component-vm.md §Modeled variant.
 */
import { ComponentVMBase } from "./componentVMBase.js";
import { ViewModelType } from "./types.js";
import type { IMessageHub } from "../services/messageHub.js";
import type { IDispatcher } from "../services/dispatcher.js";
import { NullMessageHub } from "../services/nullMessageHub.js";
import { NullDispatcher } from "../services/nullDispatcher.js";
import { BuilderValidationError } from "../builders/exceptions.js";

export class ComponentVMOf<M> extends ComponentVMBase {
  #model: M;
  readonly #modeledHinter: (m: M) => string;
  readonly #onModelChangedCb: ((m: M) => void) | null;
  #modeledHint: string;
  readonly #vmType: ViewModelType;

  constructor(opts: {
    name: string;
    hint: string;
    initialModel: M;
    modeledHinter: (m: M) => string;
    onModelChanged?: ((m: M) => void) | null;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    onConstruct?: (() => void) | null;
    onDestruct?: (() => void) | null;
    background?: boolean;
    vmType?: ViewModelType;
  }) {
    super(opts);
    this.#model = opts.initialModel;
    this.#modeledHinter = opts.modeledHinter;
    this.#onModelChangedCb = opts.onModelChanged ?? null;
    this.#modeledHint = opts.modeledHinter(opts.initialModel);
    this.#vmType = opts.vmType ?? ViewModelType.Component;
  }

  get type(): ViewModelType {
    return this.#vmType;
  }

  get model(): M {
    return this.#model;
  }

  set model(value: M) {
    this._setModel(value);
  }

  get modeledHint(): string {
    return this.#modeledHint;
  }

  protected _setModel(value: M): void {
    if (this.#model === value) return;
    this.#model = value;

    this._notifyPropertyChanged("model");

    const newHint = this.#modeledHinter(value);
    if (this.#modeledHint !== newHint) {
      this.#modeledHint = newHint;
      this._notifyPropertyChanged("modeledHint");
    }

    if (this.#onModelChangedCb !== null) {
      this.#onModelChangedCb(value);
    }
  }

  static builder<M>(): ComponentVMOfBuilder<M> {
    return new ComponentVMOfBuilder<M>();
  }

  /**
   * Constructs a {@link ComponentVMOf} from an options object in a single call —
   * an additive alternative to the fluent {@link ComponentVMOfBuilder}.
   * Delegates to that builder, so the required-field validation
   * ({@link BuilderValidationError} on a missing name/services) and the
   * resulting VM are identical to the fluent path.
   */
  static create<M>(options: ComponentVMOfOptions<M>): ComponentVMOf<M> {
    // Widen to Partial so the required-field guards remain meaningful for JS
    // callers / casts that bypass the type; validation is delegated to build().
    const o = options as Partial<ComponentVMOfOptions<M>>;
    let b = new ComponentVMOfBuilder<M>().model(options.model);
    if (o.name !== undefined) b = b.name(o.name);
    if (o.hint !== undefined) b = b.hint(o.hint);
    if (o.hub !== undefined && o.dispatcher !== undefined) b = b.services(o.hub, o.dispatcher);
    if (o.modeledHinter !== undefined) b = b.modeledHinter(o.modeledHinter);
    if (o.onModelChanged !== undefined) b = b.onModelChanged(o.onModelChanged);
    if (o.onConstruct !== undefined) b = b.onConstruct(o.onConstruct);
    if (o.onDestruct !== undefined) b = b.onDestruct(o.onDestruct);
    if (o.background !== undefined) b = b.background(o.background);
    if (o.vmType !== undefined) b = b.vmType(o.vmType);
    return b.build();
  }
}

/**
 * Options for the additive {@link ComponentVMOf.create} construction form
 * (ADR-0055 / VMX-020). A one-call alternative to the fluent
 * {@link ComponentVMOfBuilder}.
 */
export interface ComponentVMOfOptions<M> {
  /** Required VM name. */
  name: string;
  /** Required message hub. */
  hub: IMessageHub;
  /** Required dispatcher. */
  dispatcher: IDispatcher;
  /** Required model value. */
  model: M;
  /** Optional hint (default: ""). */
  hint?: string;
  /** Optional modeled-hint projection (default: () => ""). */
  modeledHinter?: (m: M) => string;
  /** Optional OnModelChanged callback. */
  onModelChanged?: (m: M) => void;
  /** Optional OnConstruct lifecycle callback. */
  onConstruct?: () => void;
  /** Optional OnDestruct lifecycle callback. */
  onDestruct?: () => void;
  /** Optional background-construction flag (default: false). */
  background?: boolean;
  /** Optional VM type (default: ViewModelType.Component). */
  vmType?: ViewModelType;
}

// ---------------------------------------------------------------------------
// Builder
// ---------------------------------------------------------------------------

const SENTINEL = Symbol("not-set");

export class ComponentVMOfBuilder<M> {
  #name: string | null = null;
  #hint = "";
  #model: M | typeof SENTINEL = SENTINEL;
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #modeledHinter: ((m: M) => string) | null = null;
  #onModelChanged: ((m: M) => void) | null = null;
  #onConstruct: (() => void) | null = null;
  #onDestruct: (() => void) | null = null;
  #background = false;
  #vmType: ViewModelType = ViewModelType.Component;

  constructor(from?: ComponentVMOfBuilder<M>) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#model = from.#model;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#modeledHinter = from.#modeledHinter;
      this.#onModelChanged = from.#onModelChanged;
      this.#onConstruct = from.#onConstruct;
      this.#onDestruct = from.#onDestruct;
      this.#background = from.#background;
      this.#vmType = from.#vmType;
    }
  }

  name(value: string): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#name = value;
    return b;
  }

  hint(value: string): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#hint = value;
    return b;
  }

  model(value: M): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#model = value;
    return b;
  }

  services(hub: IMessageHub, dispatcher: IDispatcher): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  modeledHinter(fn: (m: M) => string): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#modeledHinter = fn;
    return b;
  }

  onModelChanged(cb: (m: M) => void): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#onModelChanged = cb;
    return b;
  }

  onConstruct(cb: () => void): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#onConstruct = cb;
    return b;
  }

  onDestruct(cb: () => void): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#onDestruct = cb;
    return b;
  }

  background(value: boolean): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#background = value;
    return b;
  }

  vmType(value: ViewModelType): ComponentVMOfBuilder<M> {
    const b = new ComponentVMOfBuilder<M>(this);
    b.#vmType = value;
    return b;
  }

  /**
   * Chainable Wither that wires {@link NullMessageHub.INSTANCE} +
   * {@link NullDispatcher.INSTANCE} in a single call. Mirrors C#'s
   * `WithNullServices<M>()` and Python's `with_null_services()` per
   * spec/10-builders.md / ADR-0035.
   */
  withNullServices(): ComponentVMOfBuilder<M> {
    return this.services(NullMessageHub.INSTANCE, NullDispatcher.INSTANCE);
  }

  build(): ComponentVMOf<M> {
    if (this.#name === null) throw new BuilderValidationError("name");
    if (this.#model === SENTINEL) throw new BuilderValidationError("model");
    if (this.#hub === null || this.#dispatcher === null)
      throw new BuilderValidationError("services");

    const hinter = this.#modeledHinter ?? ((): string => "");
    const initialModel = this.#model;
    return new ComponentVMOf<M>({
      name: this.#name,
      hint: this.#hint,
      initialModel,
      modeledHinter: hinter,
      onModelChanged: this.#onModelChanged,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      onConstruct: this.#onConstruct,
      onDestruct: this.#onDestruct,
      background: this.#background,
      vmType: this.#vmType,
    });
  }
}

File: langs/swift/Sources/VMx/Components/ComponentVMOf.swift
//
// ComponentVMOf<Model> — modeled leaf viewmodel with a settable model.
//
// Mirrors C#'s `ComponentVM<M>` and TS / Python `ComponentVMOf<M>`.
//
// See spec/05-component-vm.md §Modeled variant.
//
import Foundation

/// Default `modelEquals` predicate for **non-`Equatable`** models.
///
/// HUB-005 requires that setting `model` to a value equal to the current one
/// publishes nothing. `Equatable` models get structural `==` (via the
/// constrained `init` / `build()` overloads). A non-`Equatable` model can't be
/// compared structurally, but if it is a **reference type (class)** we can
/// still suppress the redundant publish when the new model is the *same
/// instance* — matching the reference-identity suppression the C# / Python / TS
/// flavors apply. Non-`Equatable` **value** models (a `struct`/`enum` without
/// an `Equatable` conformance) fall through to "changed" so every set still
/// publishes — pass an explicit `.modelEquals(_:)` for structural suppression.
///
/// `@usableFromInline` so the public `ComponentVMOf.init` can name it as a
/// default-argument value (default args are emitted into the caller's module).
@usableFromInline
func _defaultModelEquals<Model>(_ lhs: Model, _ rhs: Model) -> Bool {
    // Only apply reference identity for actual class instances. A value type
    // would bridge to a freshly-boxed `AnyObject` here (or, for `Int` & friends,
    // to a tagged `NSNumber`), so gating on `is AnyClass` keeps value-model
    // behaviour unchanged — they report "not equal" and publish as before.
    guard type(of: lhs) is AnyClass else { return false }
    return (lhs as AnyObject) === (rhs as AnyObject)
}

public struct ComponentVMOfOptions<Model> {
    public var name: String?
    public var hint: String
    public var model: Model
    public var hub: MessageHubProtocol?
    public var dispatcher: Dispatcher?
    public var modeledHinter: ((Model) -> String)?
    public var modelEquals: ((Model, Model) -> Bool)?
    public var onModelChanged: ((Model) -> Void)?
    public var onConstruct: (() -> Void)?
    public var onDestruct: (() -> Void)?
    public var background: Bool

    public init(
        name: String? = nil,
        hint: String = "",
        model: Model,
        hub: MessageHubProtocol? = nil,
        dispatcher: Dispatcher? = nil,
        modeledHinter: ((Model) -> String)? = nil,
        modelEquals: ((Model, Model) -> Bool)? = nil,
        onModelChanged: ((Model) -> Void)? = nil,
        onConstruct: (() -> Void)? = nil,
        onDestruct: (() -> Void)? = nil,
        background: Bool = false
    ) {
        self.name = name
        self.hint = hint
        self.model = model
        self.hub = hub
        self.dispatcher = dispatcher
        self.modeledHinter = modeledHinter
        self.modelEquals = modelEquals
        self.onModelChanged = onModelChanged
        self.onConstruct = onConstruct
        self.onDestruct = onDestruct
        self.background = background
    }
}

open class ComponentVMOf<Model>: ComponentVMBase {
    private var _model: Model
    private let modeledHinter: (Model) -> String
    private let onModelChangedCb: ((Model) -> Void)?
    private var _modeledHint: String
    private let modelEquals: (Model, Model) -> Bool

    public init(
        name: String,
        hint: String = "",
        initialModel: Model,
        modeledHinter: @escaping (Model) -> String = { _ in "" },
        modelEquals: @escaping (Model, Model) -> Bool = { _defaultModelEquals($0, $1) },
        onModelChanged: ((Model) -> Void)? = nil,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        onConstruct: (() -> Void)? = nil,
        onDestruct: (() -> Void)? = nil,
        background: Bool = false
    ) {
        self._model = initialModel
        self.modeledHinter = modeledHinter
        self.modelEquals = modelEquals
        self.onModelChangedCb = onModelChanged
        self._modeledHint = modeledHinter(initialModel)
        super.init(
            name: name,
            hint: hint,
            hub: hub,
            dispatcher: dispatcher,
            onConstruct: onConstruct,
            onDestruct: onDestruct,
            background: background
        )
    }

    open override var type: ViewModelType { .component }

    public var model: Model {
        get { _model }
        set { _setModel(newValue) }
    }

    public var modeledHint: String { _modeledHint }

    /// Internal setter so subclasses (e.g. `ReadonlyComponentVMOf`) can
    /// gate writes while still using the same machinery.
    func _setModel(_ value: Model) {
        if modelEquals(_model, value) { return }
        _model = value
        // Inert-disposed-VM principle extended to the model setter: a disposed VM
        // publishes nothing further. NOTE spec/02 invariant 3 specifies this only
        // for `IsCurrent` selection, not `model`, so model-set-after-dispose is
        // UNSPECIFIED and divergent cross-flavor — C#/Python/TS publish here
        // (ADR-0009). The model *field* is still updated above (so the getter
        // reflects the latest value — parity with C#, whose field write is
        // unconditional); only the hub send / raise / hint / callback is gated.
        guard status != .disposed else { return }

        hub.send(PropertyChangedMessage(
            sender: self, senderName: name, propertyName: "model"
        ))
        _raisePropertyChanged("model")

        let newHint = modeledHinter(value)
        if newHint != _modeledHint {
            _modeledHint = newHint
            hub.send(PropertyChangedMessage(
                sender: self, senderName: name, propertyName: "modeledHint"
            ))
            _raisePropertyChanged("modeledHint")
        }

        onModelChangedCb?(value)
    }

    public static func builder() -> ComponentVMOfBuilder<Model> {
        ComponentVMOfBuilder<Model>()
    }

    public static func create(_ options: ComponentVMOfOptions<Model>) throws -> ComponentVMOf<Model> {
        var b = ComponentVMOf<Model>.builder()
            .hint(options.hint)
            .model(options.model)
            .background(options.background)
        if let name = options.name { b = b.name(name) }
        if let hub = options.hub, let dispatcher = options.dispatcher {
            b = b.services(hub: hub, dispatcher: dispatcher)
        }
        if let modeledHinter = options.modeledHinter { b = b.modeledHinter(modeledHinter) }
        if let modelEquals = options.modelEquals { b = b.modelEquals(modelEquals) }
        if let onModelChanged = options.onModelChanged { b = b.onModelChanged(onModelChanged) }
        if let onConstruct = options.onConstruct { b = b.onConstruct(onConstruct) }
        if let onDestruct = options.onDestruct { b = b.onDestruct(onDestruct) }
        return try b.build()
    }
}

// ─── Equatable convenience overload ──────────────────────────────────────
//
// When the model conforms to `Equatable`, `ComponentVMOf<M>.builder()`
// callers don't need to pass an explicit `.modelEquals(...)` predicate —
// the builder defaults to `==`. See `ComponentVMOfBuilder` for details.

extension ComponentVMOf where Model: Equatable {
    public static func create(_ options: ComponentVMOfOptions<Model>) throws -> ComponentVMOf<Model> {
        var b = ComponentVMOf<Model>.builder()
            .hint(options.hint)
            .model(options.model)
            .background(options.background)
        if let name = options.name { b = b.name(name) }
        if let hub = options.hub, let dispatcher = options.dispatcher {
            b = b.services(hub: hub, dispatcher: dispatcher)
        }
        if let modeledHinter = options.modeledHinter { b = b.modeledHinter(modeledHinter) }
        if let modelEquals = options.modelEquals { b = b.modelEquals(modelEquals) }
        if let onModelChanged = options.onModelChanged { b = b.onModelChanged(onModelChanged) }
        if let onConstruct = options.onConstruct { b = b.onConstruct(onConstruct) }
        if let onDestruct = options.onDestruct { b = b.onDestruct(onDestruct) }
        return try b.build()
    }

    /// Equatable-aware initializer convenience — defaults `modelEquals`
    /// to `==`. (The builder path gets the same default via the
    /// constraint-overloaded `ComponentVMOfBuilder.build()` — a static
    /// `builder()` overload here made annotation-free call sites
    /// ambiguous, which CI caught.)
    public convenience init(
        name: String,
        hint: String = "",
        initialModel: Model,
        modeledHinter: @escaping (Model) -> String = { _ in "" },
        onModelChanged: ((Model) -> Void)? = nil,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        onConstruct: (() -> Void)? = nil,
        onDestruct: (() -> Void)? = nil,
        background: Bool = false
    ) {
        self.init(
            name: name,
            hint: hint,
            initialModel: initialModel,
            modeledHinter: modeledHinter,
            modelEquals: ==,
            onModelChanged: onModelChanged,
            hub: hub,
            dispatcher: dispatcher,
            onConstruct: onConstruct,
            onDestruct: onDestruct,
            background: background
        )
    }
}
